ldon/src/store.rs

```diff
@@ -0,0 +1,658 @@
+    writeln!(f, "{{")?;
+    for (k, v) in self.0.iter() {
+      match v {
+        Entry::Expr(x) => {
+          writeln!(f, "  {}: {},", k, x)?;
+        },
+        Entry::Cont(x) => {
+          writeln!(f, "  {}: {:?},", k, x)?;
+        },
+        Entry::Opaque => {
+          writeln!(f, "  {}: _,", k)?;
+        },
+      }
+    }
+    writeln!(f, "}}")?;
+    Ok(())
+  }
+}
+
+impl<F: LurkField> SerdeF<F> for Store<F> {
+  fn ser_f(&self) -> Vec<F> {
+    let mut exprs = Vec::new();
+    let mut opaqs = Vec::new();
+    for (ptr, entry) in self.0.iter() {
+      match entry {
+        Entry::Expr(x) => exprs.extend(x.ser_f().into_iter()),
+        Entry::Cont(_x) => todo!(),
+        Entry::Opaque => opaqs.extend(ptr.ser_f()),
+      }
+    }
+    let mut res = vec![(opaqs.len() as u64).into()];
+    res.extend(opaqs);
+    res.extend(exprs);
+    res
+  }
+
+  fn de_f(fs: &[F]) -> Result<Store<F>, SerdeFError<F>> {
+    let mut map: BTreeMap<Ptr<F>, Entry<F>> = BTreeMap::new();
+    if fs.is_empty() {
+      return Err(SerdeFError::UnexpectedEnd);
+    }
+    let opaqs: u64 =
+      fs[0].to_u64().ok_or_else(|| SerdeFError::ExpectedU64(fs[0]))?;
+    // TODO: This will break on 32-bit targets, but maybe we don't care.
+    let opaqs: usize = opaqs as usize;
+    if fs.len() < opaqs {
+      return Err(SerdeFError::UnexpectedEnd);
+    }
+    let mut i = 1;
+    while i <= opaqs {
+      map.insert(Ptr::de_f(&fs[i..])?, Entry::Opaque);
+      i += 2;
+    }
+    while i < fs.len() {
+      let ptr = Ptr::de_f(&fs[i..])?;
+      match ptr.tag.kind {
+        TagKind::Expr(_) => {
+          let expr = Expr::de_f(&fs[i..])?;
+          map.insert(ptr, Entry::Expr(expr));
+          i += 2 + expr.child_ptrs().len() * 2;
+        },
+        TagKind::Cont(_) => todo!(),
+        TagKind::Op1(_) => todo!(),
+        TagKind::Op2(_) => todo!(),
+      }
+    }
+    Ok(Store(map))
+  }
+}
+#[cfg(feature = "test-utils")]
+pub mod test_utils {
+  use blstrs::Scalar as Fr;
+  use lurk_ff::test_utils::frequency;
+  use quickcheck::{
+    Arbitrary,
+    Gen,
+  };
+
+  use super::*;
+  impl Arbitrary for Entry<Fr> {
+    fn arbitrary(g: &mut Gen) -> Self {
+      #[allow(clippy::type_complexity)]
+      let input: Vec<(i64, Box<dyn Fn(&mut Gen) -> Entry<Fr>>)> = vec![
+        (100, Box::new(|_| Self::Opaque)),
+        (100, Box::new(|g| Self::Expr(Expr::arbitrary(g)))),
+      ];
+      frequency(g, input)
+    }
+  }
+
+  impl Arbitrary for Store<Fr> {
+    fn arbitrary(g: &mut Gen) -> Self {
+      let mut map: BTreeMap<Ptr<Fr>, Entry<Fr>> = BTreeMap::new();
+      let n: usize = usize::arbitrary(g) % 5;
+      let cache = PoseidonCache::default();
+      for _ in 0..n {
+        let entry = Entry::arbitrary(g);
+        match entry {
+          Entry::Opaque => map.insert(Ptr::arbitrary(g), entry),
+          Entry::Expr(x) => map.insert(x.ptr(&cache), entry),
+          Entry::Cont(_) => todo!(),
```

Comment:
We'll want to have `Store` passing property tests before merging

Reply:
Okay, I will implement serialization for `Cont` to get this working